aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -501,6 +501,119 @@ Tensor norm(const Tensor& self, Scalar p) {
   return at::native::_norm(self, p);
 }
 
+static Tensor& _norm_matrix_out(Tensor &result, const Tensor &self, optional<Scalar> opt_p,
+                               IntArrayRef dim, bool keepdim, optional<ScalarType> opt_dtype) {
+  auto p = opt_p.value_or(2.0).toDouble();
+  TORCH_CHECK(self.device().type() == DeviceType::CPU || self.device().type() == DeviceType::CUDA,
+              "_norm_matrix only supports CPU AND CUDA device type, got: ", self.device().type());
+  TORCH_CHECK(self.layout() == Layout::Strided,
+              "_norm_matrix only supports strided layout, got: ", self.layout());
+  if ((dim.size() == 0) && (self.dim() == 2)) {
+    dim = {0, 1};
+  }
+  TORCH_CHECK(dim.size() == 2, "_norm_matrix: 'dim' must either specify 2 dimensions, or if ",
+    "'self' is 2-D 'dim' can specify 0 dimensions for a full reduction. Got 'dim' specifying ",
+    dim.size(), " dims and 'self' is ", self.dim(), "-D");
+  ScalarType scalarType = opt_dtype.has_value() ? opt_dtype.value() : self.scalar_type();
+  TORCH_CHECK(
+      at::isFloatingType(scalarType) || at::isComplexType(scalarType),
+      "Can only calculate the mean of floating types. Got ",
+      toString(scalarType),
+      " instead.");
+
+  auto dim_ = dim.vec();
+  maybe_wrap_dims(dim_, self.dim());
+  TORCH_CHECK(dim_[0] != dim_[1],
+    "Expected dims to be different, got (", dim[0], ", ", dim[1], ") instead");
+
+  Tensor result_;
+  Tensor self_;
+    
+  if (opt_dtype.has_value()) {
+    self_ = self.to(scalarType);
+  } else {
+    self_ = self;
+  }
+
+  if (std::abs(p) == 2) {
+    // Need to shift the reduction dims to the back, because at::svd will only operate on
+    // the last 2 dimensions
+    std::vector<int64_t> permuted_dims(self_.dim());
+    int64_t cur_permuted_dim = 0;
+    for (int64_t dim_ind = 0; dim_ind < self_.dim(); dim_ind++) {
+      if ((dim_ind != dim_[0]) && (dim_ind != dim_[1])) {
+        permuted_dims[cur_permuted_dim++] = dim_ind;
+      }
+    }
+    permuted_dims[cur_permuted_dim++] = dim_[0];
+    permuted_dims[cur_permuted_dim] = dim_[1];
+
+    result_ = std::get<1>(self_.permute(permuted_dims).svd()).abs();
+
+    if (p > 0) {
+      result_ = std::get<0>(result_.max(-1, keepdim));
+    } else {
+      result_ = std::get<0>(result_.min(-1, keepdim));
+    }
+  } else {
+    if (std::abs(p) == INFINITY) {
+      std::swap(dim_[0], dim_[1]);
+    }
+    if ((dim_[1] > dim_[0]) && !keepdim) {
+      dim_[1]--;
+    }
+    if (p == 1 || p == INFINITY) {
+      result_ = std::get<0>(self_.abs().sum(dim_[0], keepdim).max(dim_[1], keepdim));
+    } else if (p == -1 || p == -INFINITY) {
+      result_ = std::get<0>(self_.abs().sum(dim_[0], keepdim).min(dim_[1], keepdim));
+    } else {
```

Comment:
The code with this dim swap looks very convoluted to me. Something like this is way easier to understand:
```
if p == 1 return input.abs().sum(-2).max(-1)
elif p == -1 return input.abs().sum(-2).min(-1)
elif p == infty return input.abs().sum(-1).max(-1)
elif p == -infty return input.abs().sum(-1).max(-1)
```

Reply:
That's fair. I'll make that change.

Reply:
Although, changing them to 4 separate cases will make the code required for the `if (second_reduction_dim > first_reduction_dim) { second_reduction_dim--; }` correction a bit more complicated. I would argue that conditionally swapping the dims makes accomplishing that correction simpler, and that if we don't swap dims, the logic as a whole will have to be more complicated.

Reply:
Aaah, sorry, dim is an array fed as a parameter. My bad...

Reply:
Maybe to make the logic more clear, I could do something like this:
```
int64_t first_reduction_dim = (std::abs(p) == INFINITY) ? dim_[1] : dim[0];
int64_t second_reduction_dim = (std::abs(p) == INFINITY) ? dim_[0] : dim[1];
```

Reply:
> Although, changing them to 4 separate cases will make the code required for the `if (second_reduction_dim > first_reduction_dim) { second_reduction_dim--; }` correction a bit more complicated. I would argue that conditionally swapping the dims makes accomplishing that correction simpler, and that if we don't swap dims, the logic as a whole will have to be more complicated.

Could you please explain why you need this decrement?

Reply:
Maybe also putting things into `switch` will make things easier to grasp?

Reply:
If I have `dim = (0, 1)` and `a.dim() == 2`, then I do `b = a.sum(dim[0])`, b will have one fewer dimension. So the subsequent `b.max(dim[1])` will fail because b doesn't have dimension 1.

Reply:
Could you please put a comment about it? It is not obvious. And, probably, you do not this if `keepdim=True`.

Reply:
Sure thing